Calculate scroll position from the bottom

knubie opened the pull request:
Previously, when scrolling to the top to fetch messages, the new scroll
position would be calculated based on the number of messages and their
heights.

With this commit, the new scroll position is based on the distance from
the bottom before messages are prepended, since that value should be
equal before and after.

E.g. if the scroll position is 1000px from the bottom before the new
messages are prepended, we make sure the new scroll position is also
1000px from the bottom after they're prepended.

fixes #18

knubie committed:
Calculate scroll position from the bottom

Previously, when scrolling to the top to fetch messages, the new scroll
position would be calculated based on the number of messages and their
heights.

With this commit, the new scroll position is based on the distance from
the bottom before messages are prepended, since that value should be
equal before and after.

E.g. if the scroll position is 1000px from the bottom before the new
messages are prepended, we make sure the new scroll position is also
1000px from the bottom after they're prepended.

galenwp commented:
The code for this always felt too bloated.  Did you also test this with new messages being appended?

knubie commented:
Yeah this does work for messages appending to the bottom, when you're scrolled to the bottom. I.e. you're at scrollBottom 0, a message comes in, you remain at scrollBottom 0. The little bit of testing I did last night seemed to confirm that.

The only scenario I can think of where this doesn't work (that I just thought of this morning) is if you are scrolled up to the middle somewhere, for example reading the backlog, and new messages come in. It would scroll you down a bit for each message. In that scenario you would actually want the scrollTop to be the same, not scrollBottom.

In that case you might add a conditional that looks like

``` coffeescript
componentDidUpdate: ->
  if messageAppendedToBottom && @ScrollbBottom > 0
    window.scrollTop @prevScrollTop
```

knubie committed:
Prevent appended messages scrolling

This prevents incoming messages appended to the bottom from scrolling
the message list down if the user is not scrolled all the way to the
bottom.

galenwp commented:
And you got it!  Thanks.  That was already an issue, actually.  @juped will be really happy to see this fixed.

galenwp commented:
Want to try rolling this into a https://github.com/urbit/arvo PR too?

knubie commented:
Sure! What is the protocol there, just moving the `main.css` and `main.js` files over and referencing the original commit?

galenwp commented:
Yep.  this is just a `main.js` change — so you can just add that.  Then the commit should be something like `talk@d80a6c7`.

Thanks!
